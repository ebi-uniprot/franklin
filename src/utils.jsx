import React from 'react';

export function getLastIndexOfSubstringIgnoreCase(string, substring) {
  return string.toLowerCase().lastIndexOf(substring.toLowerCase());
}

export const getFlattenedPaths = (currentItems, id, path = []) => {
  let flattened = [];
  currentItems.forEach((node) => {
    const { items, ...thisNode } = node;
    const nodePath = [...path, thisNode];
    if (items) {
      const result = getFlattenedPaths(items, id, nodePath);
      if (result.length) {
        flattened = [...flattened, ...result];
      }
    } else if (!id || thisNode.id === id) {
      flattened = [...flattened, nodePath];
    }
  });
  return flattened;
};

export function restructureFlattenedTreeItemsForAutocomplete(
  items,
  sep = ' / '
) {
  return {
    id: items[items.length - 1].id,
    pathLabel: items.map((item) => item.label).join(sep),
    itemLabel: items[items.length - 1].label,
  };
}

export function restructureFlattenedTreeDataForAutocomplete(flattenedTreeData) {
  return flattenedTreeData.map((items) =>
    restructureFlattenedTreeItemsForAutocomplete(items)
  );
}

export function formatLargeNumber(x) {
  return x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

export function capitaliseFirstLetter(string) {
  return string && string.charAt(0).toUpperCase() + string.slice(1);
}

export function highlightSubstring(string, substring) {
  const i = getLastIndexOfSubstringIgnoreCase(string, substring);
  if (i < 0) return string;
  const prestring = string.slice(0, i);
  const highlight = string.slice(i, i + substring.length);
  const poststring = string.slice(i + substring.length);
  return (
    <>
      {prestring}
      <b>{highlight}</b>
      {poststring}
    </>
  );
}

export function getClassNames(mixedInput, optionalCondition, output = '') {
  if (typeof optionalCondition !== 'undefined' && optionalCondition !== true) {
    return '';
  }

  if (typeof mixedInput === 'string') {
    return mixedInput;
  }

  if (Array.isArray(mixedInput)) {
    if (typeof mixedInput[1] === 'boolean') {
      return getClassNames(mixedInput[0], mixedInput[1], output);
    }

    return mixedInput
      .map((item) => getClassNames(item, optionalCondition, output))
      .filter((x) => x) // removes nulls, undefineds, empty strings, etc.
      .join(' ');
  }

  return output;
}
