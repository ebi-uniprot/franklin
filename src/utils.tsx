export function getLastIndexOfSubstringIgnoreCase(
  string: string,
  substring: string
) {
  return string.toLowerCase().lastIndexOf(substring.toLowerCase());
}

export type Item = {
  label: string;
  id: string;
  items?: Item[];
};

export const getFlattenedPaths = (
  currentItems: Item[],
  id?: string,
  path: Item[] = []
) => {
  let flattened: Pick<Item, 'label' | 'id'>[][] = [];
  currentItems.forEach((node) => {
    const { items, ...thisNode } = node;
    const nodePath = [...path, thisNode];
    if (items) {
      const result = getFlattenedPaths(items, id, nodePath);
      if (result.length) {
        flattened = [...flattened, ...result];
      }
    } else if (!id || thisNode.id === id) {
      flattened = [...flattened, nodePath];
    }
  });
  return flattened;
};

export function restructureFlattenedTreeItemsForAutocomplete(
  items: Item[],
  sep = ' / '
) {
  return {
    id: items[items.length - 1].id,
    pathLabel: items.map((item) => item.label).join(sep),
    itemLabel: items[items.length - 1].label,
  };
}

export function restructureFlattenedTreeDataForAutocomplete(
  flattenedTreeData: Item[][]
) {
  return flattenedTreeData.map((items) =>
    restructureFlattenedTreeItemsForAutocomplete(items)
  );
}

export function formatLargeNumber(x: string | number) {
  return x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

export function capitaliseFirstLetter(string?: string | null) {
  return string && string.charAt(0).toUpperCase() + string.slice(1);
}

export function highlightSubstring(string: string, substring: string) {
  const i = getLastIndexOfSubstringIgnoreCase(string, substring);
  if (i < 0) return string;
  const prestring = string.slice(0, i);
  const highlight = string.slice(i, i + substring.length);
  const poststring = string.slice(i + substring.length);
  return (
    <>
      {prestring}
      <b>{highlight}</b>
      {poststring}
    </>
  );
}

const reProtocol = /^(https?:)?(\/\/)?/;
const reTrailingSlashes = /(\/+$)/;

export const tidyUrlString = (url: string) =>
  url.replace(reProtocol, '').replace(reTrailingSlashes, '');
