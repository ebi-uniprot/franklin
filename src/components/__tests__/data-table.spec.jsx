import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import DataTable from '../data-table';
import { fillArray } from '../../utils';

describe('DataTable', () => {
  const getIdKey = ({ id }) => id;
  const onSelect = jest.fn();
  const onHeaderClick = jest.fn();
  const selectable = true;
  const selected = ['id0'];
  const columns = [
    {
      label: 'Column 1',
      name: 'content1',
      render: (row) => row.content1,
      sortable: true,
      sorted: 'ascend',
    },
    {
      label: 'Column 2',
      name: 'content2',
      render: (row) => row.content2,
    },
    {
      label: 'Column 3',
      name: 'content3',
      render: (row) => row.content3,
      sortable: true,
    },
  ];
  const data = fillArray(10, (_, index) => ({
    id: `id${index}`,
    content1: 'foo',
    content2: 'bar',
    content3: 'baz',
  }));
  const scrollDataAttribute = 'data-table';

  let onLoadMoreItems;

  const getRender = (hasMoreData = true) =>
    render(
      <div
        style={{ height: '65vh', overflowY: 'auto' }}
        data-loader-scroll={scrollDataAttribute}
        data-testid="scroll-container"
      >
        <DataTable
          onLoadMoreItems={onLoadMoreItems}
          hasMoreData={hasMoreData}
          data={data}
          getIdKey={getIdKey}
          columns={columns}
          onSelect={onSelect}
          selected={selected}
          onHeaderClick={onHeaderClick}
          selectable={selectable}
          scrollDataAttribute={scrollDataAttribute}
        />
      </div>
    );

  beforeEach(() => {
    onLoadMoreItems = jest.fn();
  });

  test('should render', () => {
    const { asFragment } = getRender();
    expect(asFragment()).toMatchSnapshot();
  });

  test('should request more data', () => {
    const { getByTestId } = getRender();
    const scrollContainer = getByTestId('scroll-container');
    fireEvent.scroll(scrollContainer, { target: { scrollY: 1000 } });
    expect(onLoadMoreItems).toHaveBeenCalled();
  });

  test('should not request more data when hasMoreData is false', () => {
    const { getByTestId } = getRender(false);
    const scrollContainer = getByTestId('scroll-container');
    fireEvent.scroll(scrollContainer, { target: { scrollY: 1000 } });
    expect(onLoadMoreItems).toHaveBeenCalledTimes(0);
  });

  test('should fire onSelect when checkbox is clicked', () => {
    const { container } = getRender();
    const checkbox = container.querySelector('tbody tr td input');
    fireEvent.click(checkbox);
    expect(onSelect).toHaveBeenCalled();
  });

  test('should fire onHeaderClick when header is clicked', () => {
    const { container } = getRender();
    const sortableHeader = container.querySelector(
      '.data-table__header-cell--ascend'
    );
    fireEvent.click(sortableHeader);
    expect(onHeaderClick).toHaveBeenCalled();
  });
});
