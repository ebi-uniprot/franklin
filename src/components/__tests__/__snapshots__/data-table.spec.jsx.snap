// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable should render 1`] = `
<DocumentFragment>
  <div
    data-loader-scroll="data-table"
    data-testid="scroll-container"
    style="height: 65vh; overflow-y: auto;"
  >
    <table
      class="data-table   "
    >
      <thead
        class="data-table__header"
      >
        <tr
          class="data-table__row"
        >
          <th
            class="data-table__header-cell data-table__header-cell--checkbox"
          >
             
          </th>
          <th
            class="data-table__header-cell data-table__header-cell--ascend"
            style="width: auto;"
          >
            Column 1
          </th>
          <th
            class="data-table__header-cell "
            style="width: auto;"
          >
            Column 2
          </th>
          <th
            class="data-table__header-cell data-table__header-cell--sortable"
            style="width: auto;"
          >
            Column 3
          </th>
        </tr>
      </thead>
      <tbody
        class="data-table__table__body"
      >
        <tr>
          <td
            class="data-table__cell data-table__cell--selected"
          >
            <input
              aria-label="id0"
              checked=""
              type="checkbox"
            />
          </td>
          <td
            class="data-table__cell data-table__cell--selected "
          >
            foo
          </td>
          <td
            class="data-table__cell data-table__cell--selected "
          >
            bar
          </td>
          <td
            class="data-table__cell data-table__cell--selected "
          >
            baz
          </td>
        </tr>
        <tr>
          <td
            class="data-table__cell data-table__cell--odd "
          >
            <input
              aria-label="id1"
              type="checkbox"
            />
          </td>
          <td
            class="data-table__cell data-table__cell--odd  "
          >
            foo
          </td>
          <td
            class="data-table__cell data-table__cell--odd  "
          >
            bar
          </td>
          <td
            class="data-table__cell data-table__cell--odd  "
          >
            baz
          </td>
        </tr>
        <tr>
          <td
            class="data-table__cell "
          >
            <input
              aria-label="id2"
              type="checkbox"
            />
          </td>
          <td
            class="data-table__cell  "
          >
            foo
          </td>
          <td
            class="data-table__cell  "
          >
            bar
          </td>
          <td
            class="data-table__cell  "
          >
            baz
          </td>
        </tr>
        <tr>
          <td
            class="data-table__cell data-table__cell--odd "
          >
            <input
              aria-label="id3"
              type="checkbox"
            />
          </td>
          <td
            class="data-table__cell data-table__cell--odd  "
          >
            foo
          </td>
          <td
            class="data-table__cell data-table__cell--odd  "
          >
            bar
          </td>
          <td
            class="data-table__cell data-table__cell--odd  "
          >
            baz
          </td>
        </tr>
        <tr>
          <td
            class="data-table__cell "
          >
            <input
              aria-label="id4"
              type="checkbox"
            />
          </td>
          <td
            class="data-table__cell  "
          >
            foo
          </td>
          <td
            class="data-table__cell  "
          >
            bar
          </td>
          <td
            class="data-table__cell  "
          >
            baz
          </td>
        </tr>
        <tr>
          <td
            class="data-table__cell data-table__cell--odd "
          >
            <input
              aria-label="id5"
              type="checkbox"
            />
          </td>
          <td
            class="data-table__cell data-table__cell--odd  "
          >
            foo
          </td>
          <td
            class="data-table__cell data-table__cell--odd  "
          >
            bar
          </td>
          <td
            class="data-table__cell data-table__cell--odd  "
          >
            baz
          </td>
        </tr>
        <tr>
          <td
            class="data-table__cell "
          >
            <input
              aria-label="id6"
              type="checkbox"
            />
          </td>
          <td
            class="data-table__cell  "
          >
            foo
          </td>
          <td
            class="data-table__cell  "
          >
            bar
          </td>
          <td
            class="data-table__cell  "
          >
            baz
          </td>
        </tr>
        <tr>
          <td
            class="data-table__cell data-table__cell--odd "
          >
            <input
              aria-label="id7"
              type="checkbox"
            />
          </td>
          <td
            class="data-table__cell data-table__cell--odd  "
          >
            foo
          </td>
          <td
            class="data-table__cell data-table__cell--odd  "
          >
            bar
          </td>
          <td
            class="data-table__cell data-table__cell--odd  "
          >
            baz
          </td>
        </tr>
        <tr>
          <td
            class="data-table__cell "
          >
            <input
              aria-label="id8"
              type="checkbox"
            />
          </td>
          <td
            class="data-table__cell  "
          >
            foo
          </td>
          <td
            class="data-table__cell  "
          >
            bar
          </td>
          <td
            class="data-table__cell  "
          >
            baz
          </td>
        </tr>
        <tr>
          <td
            class="data-table__cell data-table__cell--odd "
          >
            <input
              aria-label="id9"
              type="checkbox"
            />
          </td>
          <td
            class="data-table__cell data-table__cell--odd  "
          >
            foo
          </td>
          <td
            class="data-table__cell data-table__cell--odd  "
          >
            bar
          </td>
          <td
            class="data-table__cell data-table__cell--odd  "
          >
            baz
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="data-loader__loading"
    >
      Loading...
    </div>
  </div>
</DocumentFragment>
`;
